src/meta/src/hummock/manager/mod.rs

```diff
@@ -1404,8 +1429,203 @@ where
+        versioning_guard.disable_commit_epochs = true;
+        versioning_guard.current_version.clone()
+    }
+
+    #[named]
+    async fn get_compact_task_assignment_size(&self) -> usize {
+        let compaction_guard = read_lock!(self, compaction).await;
+        compaction_guard.compact_task_assignment.len()
+    }
+
+    /// Triggers compacitons to specified compaction groups.
+    /// Returns the version deltas generated by compaction tasks
+    pub async fn trigger_compaction_deterministic(
+        &self,
+        base_version_id: HummockVersionId,
+        compaction_groups: Vec<CompactionGroupId>,
+    ) -> Result<(HummockVersionId, HummockEpoch)> {
+        tracing::info!(
+            "Trigger compaction for version {}, groups {:?}",
+            base_version_id,
+            compaction_groups
+        );
+        let old_version = self.get_current_version().await;
+        assert_eq!(
+            base_version_id, old_version.id,
+            "old_version epoch {}",
+            old_version.max_committed_epoch
+        );
+
+        if compaction_groups.is_empty() {
+            return Ok((old_version.id, old_version.max_committed_epoch));
+        }
+
+        let old_size = self.get_compact_task_assignment_size().await;
+        for compaction_group in compaction_groups {
+            self.try_send_compaction_request(compaction_group)?;
+        }
+        let (schedule_ok, task_num) = self.poll_compaction_schedule_status(old_size).await;
+        let (compaction_ok, cur_version) = self
+            .poll_compaction_tasks_status(schedule_ok, task_num, &old_version)
+            .await;
+
+        tracing::info!(
+            "Compaction schedule_ok {}, task_num {} compaction_ok {}",
+            schedule_ok,
+            task_num,
+            compaction_ok,
+        );
+
+        if !schedule_ok && !compaction_ok {
+            return Ok((old_version.id, old_version.max_committed_epoch));
+        }
+        Ok((cur_version.id, cur_version.max_committed_epoch))
+    }
+
+    /// Poll the compaction task assignment to aware whether scheduling is success.
+    /// Returns (whether scheduling is success, number of tasks)
+    async fn poll_compaction_schedule_status(&self, old_size: usize) -> (bool, usize) {
+        let poll_timeout = Duration::from_secs(2);
+        let poll_interval = Duration::from_millis(20);
+        let mut poll_duration_cnt = Duration::from_millis(0);
+        let mut new_size = self.get_compact_task_assignment_size().await;
+        let mut schedule_ok = false;
+        loop {
+            if new_size > old_size {
+                schedule_ok = true;
+                break;
+            }
+
+            if poll_duration_cnt >= poll_timeout {
+                break;
+            }
+            tokio::time::sleep(poll_interval).await;
+            poll_duration_cnt += poll_interval;
+            new_size = self.get_compact_task_assignment_size().await;
+        }
```

Comment:
Ok. Should we unlimit the concurrency of compactors? Since all the compaction resume triggers are disabled, if there are too many compaction groups in the future, the compaction scheduler will just get stuck at no idle compactor.